src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala

```diff
@@ -163,127 +161,166 @@ class MarathonSchedulerActor(
-      val res = deploy(driver, plan)
-      if (origSender != Actor.noSender) origSender ! cmd.answer
-      res
+    val res = withLockFor(ids) {
+      deploy(driver, plan)
+    }
+
+    res match {
+      case Success(_) =>
+        if (origSender != Actor.noSender) origSender ! cmd.answer
+      case Failure(e: LockingFailedException) if cmd.force =>
+        deploymentManager ! CancelConflictingDeployments(plan)
+        val cancellationHandler = context.system.scheduler.scheduleOnce(
+          cancellationTimeout,
+          self,
+          CancellationTimeoutExceeded)
+
+        context.become(awaitCancellation(plan, origSender, cancellationHandler))
+      case Failure(e: LockingFailedException) =>
+        deploymentManager.ask(RetrieveRunningDeployments)(2.seconds)
+          .mapTo[RunningDeployments]
+          .foreach {
+            case RunningDeployments(plans) =>
+              val relatedDeploymentIds: Seq[String] = plans.collect {
+                case (p, _) if p.affectedApplicationIds.intersect(plan.affectedApplicationIds).nonEmpty => p.id
+              }
+              origSender ! CommandFailed(cmd, AppLockedException(relatedDeploymentIds))
+          }
     }
   }
 
-  def deploy(driver: SchedulerDriver, plan: DeploymentPlan): Future[Unit] = {
-    deploymentRepository.store(plan).flatMap { _ =>
-      val promise = Promise[Any]()
-      val promiseActor = context.actorOf(Props(classOf[PromiseActor], promise))
-      val msg = PerformDeployment(driver, plan)
-      deploymentManager.tell(msg, promiseActor)
+  def deploy(driver: SchedulerDriver, plan: DeploymentPlan): Unit = {
+    deploymentRepository.store(plan).foreach { _ =>
+      deploymentManager ! PerformDeployment(driver, plan)
+    }
+  }
 
-      val res = promise.future.map(_ => ())
+  def deploymentSuccess(plan: DeploymentPlan): Unit = {
+    log.info(s"Deployment of ${plan.target.id} successful")
+    eventBus.publish(DeploymentSuccess(plan.id))
+    deploymentRepository.expunge(plan.id)
+  }
 
-      res andThen {
-        case Success(_) =>
-          log.info(s"Deployment of ${plan.target.id} successful")
-          eventBus.publish(DeploymentSuccess(plan.id))
-          deploymentRepository.expunge(plan.id)
-
-        case Failure(e) =>
-          log.error(e, s"Deployment of ${plan.target.id} failed")
-          plan.affectedApplicationIds.foreach(appId => taskQueue.purge(appId))
-          eventBus.publish(DeploymentFailed(plan.id))
-          if (e.isInstanceOf[DeploymentCanceledException])
-            deploymentRepository.expunge(plan.id)
-      }
-    }
+  def deploymentFailed(plan: DeploymentPlan, reason: Throwable): Unit = {
+    log.error(reason, s"Deployment of ${plan.target.id} failed")
+    plan.affectedApplicationIds.foreach(appId => taskQueue.purge(appId))
+    eventBus.publish(DeploymentFailed(plan.id))
+    if (reason.isInstanceOf[DeploymentCanceledException])
+      deploymentRepository.expunge(plan.id)
```

Comment:
Why is the deployment only expunged in this case?

As you're aware, this represents a major behavior change.  Previously deployments could only stall indefinitely, not fail.  This might be OK, but we need to make sure users can figure out that the deployment went away because of a failure, and not because it completed.  For consumers that don't subscribe to the event bus (or drop a message) this aliasing in the REST API could be problematic.